src/Compilers/Core/Portable/Syntax/AbstractWarningStateMap.cs

```diff
@@ -87,5 +87,13 @@ public int CompareTo(WarningStateMapEntry other)
                 return this.Position - other.Position;
             }
         }
+
+        protected class WarningStateMapEntryComparer : System.Collections.IComparer
```

Comment:
1) Is there a reason this protected?
2) Can we doc around why we needed to do this (ie Xamarin.iOS doesn't calling through IComparable instances on structs, or whatever the reason)?
3) Make this a singleton, and use it - no need to allocate on every usage here.